File: services/customer/internal/domain/service/order_service.go
package service

import (
	"context"

	"github.com/codejsha/bookstore-microservices/customer/internal/application/port/pb/orderpb"
	"github.com/codejsha/bookstore-microservices/customer/internal/application/usecase"
	"github.com/codejsha/bookstore-microservices/customer/internal/domain/aggregate"
	"github.com/codejsha/bookstore-microservices/customer/internal/infrastructure/adapter/protostub"
)

var _ usecase.OrderUseCase = (*orderService)(nil)

func NewOrderService(
	orderGrpcClient *protostub.OrderGrpcClient,
) usecase.OrderUseCase {
	return &orderService{
		orderServiceClient: orderGrpcClient.Client,
	}
}

type orderService struct {
	orderServiceClient orderpb.OrderServiceClient
}

func (s orderService) FindAllCustomerOrders(ctx context.Context, req *orderpb.OrderFindAllProtoReq) (int64, []*aggregate.OrderAggregate, error) {
	// find orders
	resp, err := s.orderServiceClient.FindAllOrders(ctx, req)
	if err != nil {
		return 0, nil, err
	}

	// return response
	orderAggs := make([]*aggregate.OrderAggregate, 0, len(resp.Items))
	for i, item := range resp.Items {
		orderAggs[i] = aggregate.NewOrderAggregate(item)
	}
	return resp.Total, orderAggs, nil
}

func (s orderService) FindCustomerOrder(ctx context.Context, req *orderpb.OrderFindProtoReq) (*aggregate.OrderAggregate, error) {
	// find order
	resp, err := s.orderServiceClient.FindOrder(ctx, req)
	if err != nil {
		return nil, err
	}
	return aggregate.NewOrderAggregate(resp), nil
}

File: services/admin/src/main/kotlin/com/codejsha/bookstore/admin/domain/service/OrderService.kt
package com.codejsha.bookstore.admin.domain.service

import com.codejsha.bookstore.admin.application.port.restclient.OrderClient
import com.codejsha.bookstore.admin.application.usecase.OrderUseCase
import com.codejsha.bookstore.admin.domain.model.option.OrderQueryOption
import com.codejsha.platform.shared.data.ActorContext
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service

@Service
class OrderService(
    private val orderClient: OrderClient,
) : OrderUseCase {

    override suspend fun findAllOrders(option: OrderQueryOption, pageable: Pageable, context: ActorContext) =
        orderClient.findAllOrders(option, pageable)

    override suspend fun findOrder(uid: String, context: ActorContext) = orderClient.findOrder(uid)

    override suspend fun cancelOrder(uid: String, context: ActorContext) = orderClient.cancelOrder(uid)
}
